### PR TITLE
refactor OPT 1

### PR DESCRIPTION
**Pull request by yorkerlin**

### CODE REVIEW

**Comment by yorkerlin** [comment]
@karlnapf 
Every class in the OPT framework is a sub-class of `SGObject`.

**Comment by yorkerlin** [comment]
@karlnapf 
Is `travis` down?

**Comment by karlnapf** [comment]
Nope Its just busy, I pushed some things today. Your patch is a minimal refactor right? Feel free to merge it once travisbis green

File: src/shogun/machine/gp/KLDualInferenceMethod.cpp [hunk 1]
```diff
@@ -507,14 +534,20 @@ void CKLDualInferenceMethod::update_alpha()
...
 {
-	KLDualInferenceMethodMinimizer *minimizer=dynamic_cast<KLDualInferenceMethodMinimizer*>(m_minimizer);
-	REQUIRE(minimizer,"The minimizer must be an instance of KLDualInferenceMethodMinimizer\n"); 
+	CKLDualInferenceMethodMinimizer *minimizer=dynamic_cast<CKLDualInferenceMethodMinimizer*>(m_minimizer);
+	REQUIRE(minimizer,"The minimizer must be an instance of KLDualInferenceMethoMinimizer\n"); 
         KLDualInferenceMethodCostFunction* cost_fun=new  KLDualInferenceMethodCostFunction();
 	cost_fun->set_target(this);
+	bool cleanup=false;
+
+	if(this->ref_count()>1)
```
Comment (karlnapf) [inline src/shogun/machine/gp/KLDualInferenceMethod.cpp]: this can go I guess?

File: src/shogun/machine/gp/KLInference.cpp [hunk 2]
```diff
@@ -324,6 +343,9 @@ float64_t CKLInference::optimization()
 {
         KLInferenceCostFunction *cost_fun=new KLInferenceCostFunction();
         cost_fun->set_target(this);
+	bool cleanup=false;
+	if(this->ref_count()>1)
```
Comment (karlnapf) [inline src/shogun/machine/gp/KLInference.cpp]: all of them can go I guess?

File: src/shogun/optimization/NLOPTMinimizer.h [hunk 3]
```diff
@@ -61,7 +61,7 @@ class CNLOPTMinimizer: public FirstOrderMinimizer
 	 *
 	 * @return name CNLOPTMinimizer
 	 */
-	virtual const char* get_name() const { return "CNLOPTMinimizer"; }
+	virtual const char* get_name() const { return "NLOPTMinimizer"; }
```
Comment (yorkerlin) [inline src/shogun/optimization/NLOPTMinimizer.h]: @karlnapf 
The unit test of SGObject is based on get_name.

**Comment by yorkerlin** [comment]
@karlnapf 
travis will fail when it comes to `SerializationJSON`.

**Comment by yorkerlin** [comment]
@karlnapf 
Do we plan to drop JSON?  JSON cannot support `Double-precision floating-point` :(

File: src/shogun/machine/gp/KLDualInferenceMethod.cpp (hunk 1 above)
Comment (yorkerlin) [inline src/shogun/machine/gp/KLDualInferenceMethod.cpp]: NO. let me give you an example

Comment (yorkerlin) [inline src/shogun/machine/gp/KLDualInferenceMethod.cpp]: I use this trick to deal with `circular reference`.

File: src/shogun/optimization/NLOPTMinimizer.h (hunk 3 above)
Comment (karlnapf) [inline src/shogun/optimization/NLOPTMinimizer.h]: yep!

**Comment by karlnapf** [comment]
We dont plan to drop it, but to have a whole new serialization concept. THis might help at the point it is done.
I dont understand why only some tests fail. Any hacks to fix it temporarily?

File: src/shogun/machine/gp/KLDualInferenceMethod.cpp (hunk 1 above)
Comment (karlnapf) [inline src/shogun/machine/gp/KLDualInferenceMethod.cpp]: Mmmh I see,
Not sure we would do that in  here. Where is the circular thing biting you?

**Comment by yorkerlin** [comment]
The unit test will fail if some special double value (eg, INIFITY)  is used.

**Comment by karlnapf** [comment]
unfortunately, http://stackoverflow.com/questions/13581843/php-how-to-encode-infinity-or-nan-numbers-to-json

**Comment by karlnapf** [comment]
@OXPHOS @vigsterkr I guess we can change the en/decoder to use a string or so for these values? Should not be hard with cereal?

File: src/shogun/machine/gp/KLDualInferenceMethod.cpp (hunk 1 above)
Comment (yorkerlin) [inline src/shogun/machine/gp/KLDualInferenceMethod.cpp]: Let me give you an example.
Case A: in KLDualInferenceMethod::method():
at time 1:  KLDualInferenceMethod: ref_count =0
at time 2:  cost_fun->set_target(this);   (KLDualInferenceMethod: ref_count =1)
at time 3:  SG_UNREF(cost_fun);  ((KLDualInferenceMethod: ref_count =0 and will delete KLDualInferenceMethod in a method in KLDualInferenceMethod  )



Case B: in KLDualInferenceMethod::method():
at time 1:  KLDualInferenceMethod: ref_count =1
at time 2:  cost_fun->set_target(this);   (KLDualInferenceMethod: ref_count =2)
at time 3:  SG_UNREF(cost_fun);  ((KLDualInferenceMethod: ref_count =1 and will NOT delete KLDualInferenceMethod in  )

Comment (yorkerlin) [inline src/shogun/machine/gp/KLDualInferenceMethod.cpp]: I have to deal with these two cases.

Comment (yorkerlin) [inline src/shogun/machine/gp/KLDualInferenceMethod.cpp]: The following code will produce these two cases.
Case A
```
int main()
{
KLDualInferenceMethod* obj=new KLDualInferenceMethod()
obj->method();
return 0;
}
```

Case B
```
void GPC::set_obj(KLDualInferenceMethod* obj)
{
SG_REF(obj)
m_obj=obj;
}
void GPC::doSomething()
{
m_obj->method();
}
```

**Comment by yorkerlin** [comment]
@karlnapf 
what is more, JSON does not support small values. (eg, 1e-20)

File: src/shogun/optimization/lbfgs/LBFGSMinimizer.cpp [hunk 4]
```diff
@@ -57,12 +56,48 @@ LBFGSMinimizer::LBFGSMinimizer(FirstOrderCostFunction *fun)
 	init();
 }
 
-void LBFGSMinimizer::init()
+void CLBFGSMinimizer::init()
 {
 	set_lbfgs_parameters();
+	m_min_step=1e-6;
```
Comment (yorkerlin) [inline src/shogun/optimization/lbfgs/LBFGSMinimizer.cpp]: a hack to deal with issue related to JSON

**Comment by yorkerlin** [comment]
travis is green now. I merge this.

**Pull request merged**